pkg/controllers/leaderworkerset_controller.go

```diff
@@ -537,51 +540,102 @@ func (r *LeaderWorkerSetReconciler) iterateReplicas(ctx context.Context, lws *le
 		// or under rebuilding, which also indicates not ready.
-		if nominatedName != sortedPods[index].Name || (!noWorkerSts && nominatedName != sortedSts[index].Name) {
-			return false
+		if nominatedName != sortedPods[idx].Name || (!noWorkerSts && nominatedName != sortedSts[idx].Name) {
+			states[idx] = replicaState{
+				ready:   false,
+				updated: false,
+			}
+			continue
 		}
 
-		podTemplateHash := revisionutils.GetRevisionKey(&sortedPods[index])
-		if !(podTemplateHash == revisionKey && podutils.PodRunningAndReady(sortedPods[index])) {
-			return false
-		}
+		leaderUpdated := revisionutils.GetRevisionKey(&sortedPods[idx]) == revisionKey
+		leaderReady := podutils.PodRunningAndReady(sortedPods[idx])
 
 		if noWorkerSts {
-			return true
+			states[idx] = replicaState{
+				ready:   leaderReady,
+				updated: leaderUpdated,
+			}
+			continue
 		}
 
-		stsTemplateHash := revisionutils.GetRevisionKey(&sortedSts[index])
-		return stsTemplateHash == revisionKey && statefulsetutils.StatefulsetReady(sortedSts[index])
+		workersUpdated := revisionutils.GetRevisionKey(&sortedSts[idx]) == revisionKey
+		workersReady := statefulsetutils.StatefulsetReady(sortedSts[idx])
+
+		states[idx] = replicaState{
+			ready:   leaderReady && workersReady,
+			updated: leaderUpdated && workersUpdated,
+		}
 	}
 
-	var skip bool
-	var continuousReadyReplicas, lwsUnreadyReplicas int32
+	return states, nil
+}
+
+func rollingUpdatePartition(states []replicaState, stsReplicas int32, rollingStep int32, currentPartition int32) int32 {
+	continuousReadyReplicas := calculateContinuousReadyReplicas(states)
 
-	for index := stsReplicas - 1; index >= 0; index-- {
-		replicaReady := processReplica(index)
-		skip = skip || !replicaReady
+	// Update up to rollingStep replicas at once.
+	var rolliingStepPartition = utils.NonZeroValue(stsReplicas - continuousReadyReplicas - rollingStep)
 
-		if replicaReady && !skip {
-			continuousReadyReplicas++
+	// rollingStepPartition calculation above disregards the state of replicas with idx<rollingStepPartition.
+	// To prevent violating the maxUnavailable, we have to account for these replicas and increase the partition if some are not ready.
+	var unavailable int32
+	for idx := 0; idx < int(rolliingStepPartition); idx++ {
+		if !states[idx].ready {
+			unavailable++
 		}
-		if !replicaReady && index < *lws.Spec.Replicas {
-			lwsUnreadyReplicas++
+	}
+	var partition = rolliingStepPartition + unavailable
+
+	// Reduce the partition if replicas are continously not ready. It is safe since updating these replicas does not impact
+	// the availability of the LWS. This is important to prevent update from getting stuck in case maxUnavailable is already violated
+	// (for example, all replicas are not ready when rolling update is started).
+	// Note that we never drop the partition below rolliingStepPartition.
+	for idx := min(partition, stsReplicas-1); idx >= rolliingStepPartition; idx-- {
```

Comment:
What's the take here of `min(partition, stsReplicas-1)` rather than using partition directly.

Reply:
This is because partition (once increased by "unavailable") can be too big and result in index out of bound errors for the states.

This could be the case if we have a lot of unhealthy replicas and a small rolling step.

It's not a problem since we take min(partition, currentPartition), but we have to consider it in the partition reduction loop.

Reply:
Thank, let's merge this then.
/lgtm
/label tide/merge-method-squash